File: ts/src/test/base/test.indexBy.ts


import ccxt from '../../../ccxt.js';
import testSharedMethods from '../Exchange/base/test.sharedMethods.js';

function testIndexBy () {

    const exchange = new ccxt.Exchange ({
        'id': 'sampleexchange',
    });

    // Test 1: Basic list of dicts with string key
    const input1 = [
        { 'id': 'a', 'val': 1 },
        { 'id': 'b', 'val': 2 },
        { 'id': 'c', 'val': 3 },
    ];
    const expected1 = {
        'a': { 'id': 'a', 'val': 1 },
        'b': { 'id': 'b', 'val': 2 },
        'c': { 'id': 'c', 'val': 3 },
    };
    const result1 = exchange.indexBy (input1, 'id');
    testSharedMethods.assertDeepEqual (exchange, undefined, 'testIndexBy', result1, expected1);

    // Test 2: Skip elements missing the key
    const input2 = [
        { 'id': 'a', 'val': 1 },
        { 'val': 2 },
        { 'id': 'b', 'val': 3 },
    ];
    const expected2 = {
        'a': { 'id': 'a', 'val': 1 },
        'b': { 'id': 'b', 'val': 3 },
    };
    const result2 = exchange.indexBy (input2, 'id');
    testSharedMethods.assertDeepEqual (exchange, undefined, 'testIndexBy', result2, expected2);

    // Test 3: Empty array
    const input3: any[] = [];
    const expected3 = {};
    const result3 = exchange.indexBy (input3, 'id');
    testSharedMethods.assertDeepEqual (exchange, undefined, 'testIndexBy', result3, expected3);

    // Test 4: Duplicate keys (last one wins)
    const input4 = [
        { 'id': 'a', 'val': 1 },
        { 'id': 'a', 'val': 2 },
        { 'id': 'a', 'val': 3 },
    ];
    const expected4 = {
        'a': { 'id': 'a', 'val': 3 },
    };
    const result4 = exchange.indexBy (input4, 'id');
    testSharedMethods.assertDeepEqual (exchange, undefined, 'testIndexBy', result4, expected4);

    // Test 5: Numeric key values
    const input5 = [
        { 'code': 1, 'name': 'one' },
        { 'code': 2, 'name': 'two' },
        { 'code': 3, 'name': 'three' },
    ];
    const expected5 = {
        '1': { 'code': 1, 'name': 'one' },
        '2': { 'code': 2, 'name': 'two' },
        '3': { 'code': 3, 'name': 'three' },
    };
    const result5 = exchange.indexBy (input5, 'code');
    testSharedMethods.assertDeepEqual (exchange, undefined, 'testIndexBy', result5, expected5);

    // Test 6: List of arrays with integer key
    const input6 = [
        [ 'a', 1 ],
        [ 'b', 2 ],
        [ 'c', 3 ],
    ];
    const expected6 = {
        'a': [ 'a', 1 ],
        'b': [ 'b', 2 ],
        'c': [ 'c', 3 ],
    };
    const result6 = exchange.indexBy (input6, 0);
    testSharedMethods.assertDeepEqual (exchange, undefined, 'testIndexBy', result6, expected6);

    // Test 7: Single element
    const input7 = [
        { 'id': 'only', 'val': 42 },
    ];
    const expected7 = {
        'only': { 'id': 'only', 'val': 42 },
    };
    const result7 = exchange.indexBy (input7, 'id');
    testSharedMethods.assertDeepEqual (exchange, undefined, 'testIndexBy', result7, expected7);
}

export default testIndexBy;

File: julia/Ccxt/test/base/test.indexBy.jl
using Test
using Ccxt
function testIndexBy()

    exchange = get(ccxt, Symbol("Exchange"), nothing)(Dict{Symbol, Any}(
        Symbol("id") => "sampleexchange"
    ));
    input1 = [Dict{Symbol, Any}(
        Symbol("id") => "a",
        Symbol("val") => 1
    ), Dict{Symbol, Any}(
        Symbol("id") => "b",
        Symbol("val") => 2
    ), Dict{Symbol, Any}(
        Symbol("id") => "c",
        Symbol("val") => 3
    )];
    expected1 = Dict{Symbol, Any}(
        Symbol("a") => Dict{Symbol, Any}(
            Symbol("id") => "a",
            Symbol("val") => 1
        ),
        Symbol("b") => Dict{Symbol, Any}(
            Symbol("id") => "b",
            Symbol("val") => 2
        ),
        Symbol("c") => Dict{Symbol, Any}(
            Symbol("id") => "c",
            Symbol("val") => 3
        )
    );
    result1 = indexBy(exchange, input1, "id");
    assertDeepEqual(testSharedMethods, exchange, nothing, "testIndexBy", result1, expected1);
    input2 = [Dict{Symbol, Any}(
        Symbol("id") => "a",
        Symbol("val") => 1
    ), Dict{Symbol, Any}(
        Symbol("id") => "b",
        Symbol("val") => 3
    )];
    expected2 = Dict{Symbol, Any}(
        Symbol("a") => Dict{Symbol, Any}(
            Symbol("id") => "a",
            Symbol("val") => 1
        ),
        Symbol("b") => Dict{Symbol, Any}(
            Symbol("id") => "b",
            Symbol("val") => 3
        )
    );
    result2 = indexBy(exchange, input2, "id");
    assertDeepEqual(testSharedMethods, exchange, nothing, "testIndexBy", result2, expected2);
    input3 = [Dict{Symbol, Any}(
        Symbol("id") => "a",
        Symbol("val") => 1
    ), Dict{Symbol, Any}(
        Symbol("val") => 2
    ), Dict{Symbol, Any}(
        Symbol("id") => "b",
        Symbol("val") => 3
    )];
    expected3 = Dict{Symbol, Any}(
        Symbol("a") => Dict{Symbol, Any}(
            Symbol("id") => "a",
            Symbol("val") => 1
        ),
        Symbol("b") => Dict{Symbol, Any}(
            Symbol("id") => "b",
            Symbol("val") => 3
        )
    );
    result3 = indexBy(exchange, input3, "id");
    assertDeepEqual(testSharedMethods, exchange, nothing, "testIndexBy", result3, expected3);
    input4 = [];
    expected4 = Dict{Symbol, Any}();
    result4 = indexBy(exchange, input4, "id");
    assertDeepEqual(testSharedMethods, exchange, nothing, "testIndexBy", result4, expected4);
    input5 = [Dict{Symbol, Any}(
        Symbol("id") => "a",
        Symbol("val") => 1
    ), Dict{Symbol, Any}(
        Symbol("id") => "a",
        Symbol("val") => 2
    ), Dict{Symbol, Any}(
        Symbol("id") => "a",
        Symbol("val") => 3
    )];
    expected5 = Dict{Symbol, Any}(
        Symbol("a") => Dict{Symbol, Any}(
            Symbol("id") => "a",
            Symbol("val") => 3
        )
    );
    result5 = indexBy(exchange, input5, "id");
    assertDeepEqual(testSharedMethods, exchange, nothing, "testIndexBy", result5, expected5);
    input6 = [Dict{Symbol, Any}(
        Symbol("code") => 1,
        Symbol("name") => "one"
    ), Dict{Symbol, Any}(
        Symbol("code") => 2,
        Symbol("name") => "two"
    ), Dict{Symbol, Any}(
        Symbol("code") => 3,
        Symbol("name") => "three"
    )];
    expected6 = Dict{Symbol, Any}(
        Symbol("1") => Dict{Symbol, Any}(
            Symbol("code") => 1,
            Symbol("name") => "one"
        ),
        Symbol("2") => Dict{Symbol, Any}(
            Symbol("code") => 2,
            Symbol("name") => "two"
        ),
        Symbol("3") => Dict{Symbol, Any}(
            Symbol("code") => 3,
            Symbol("name") => "three"
        )
    );
    result6 = indexBy(exchange, input6, "code");
    assertDeepEqual(testSharedMethods, exchange, nothing, "testIndexBy", result6, expected6);
    input7 = [["a", 1], ["b", 2], ["c", 3]];
    expected7 = Dict{Symbol, Any}(
        Symbol("a") => ["a", 1],
        Symbol("b") => ["b", 2],
        Symbol("c") => ["c", 3]
    );
    result7 = indexBy(exchange, input7, 0);
    assertDeepEqual(testSharedMethods, exchange, nothing, "testIndexBy", result7, expected7);
    input8 = [Dict{Symbol, Any}(
        Symbol("id") => "only",
        Symbol("val") => 42
    )];
    expected8 = Dict{Symbol, Any}(
        Symbol("only") => Dict{Symbol, Any}(
            Symbol("id") => "only",
            Symbol("val") => 42
        )
    );
    result8 = indexBy(exchange, input8, "id");
    assertDeepEqual(testSharedMethods, exchange, nothing, "testIndexBy", result8, expected8);
end
